[css-counter-styles] RenderListMarker, RenderCounter should use CSSCounterStyle::text() to handle TextDirection

vitorroriz opened the pull request:
#### 8105a859925ce565fd8953e8234578b8c49ac354
<pre>
[css-counter-styles] RenderListMarker, RenderCounter should use CSSCounterStyle::text() to handle TextDirection
<a href="https://bugs.webkit.org/show_bug.cgi?id=256153">https://bugs.webkit.org/show_bug.cgi?id=256153</a>
<a href="https://rdar.apple.com/109014745">rdar://109014745</a>

Reviewed by Matthieu Dubet and Tim Nguyen.

Before we can address the bug, we need a small refactoring on how we handle
disclosure-closed counter style:
With this patch, we will handle disclosure-closed as one of the special complex systems,
such that we can consider RTL context when calculating its text value.

We fail css/css-counter-styles/counter-style-at-rule/disclosure-styles.html for 2 different
reasons. Therefore, we need to fix it in 2 different points:

1. RenderListMarker: This class is used for rendering list markers for the general case.
It has custom code for hadling TextDirection for counter styles due to disclosure-closed
style. We can now delegate this to CSSCounterStyle::text by passing TextDirection information.

The custom code currently checks if the counter-style has a &quot;disclosure-closed&quot; identifier but
that won&apos;t work for &quot;extends&quot; systems. When an author extends a counter-style, it can give any
identifier name to it, but it should still act as the counter-style it is extending.  Since this
is now handled by CSSCounterStyle itself, we no longer need the custom code.

2. RenderCounter: The class is used when rendering counters with the counter() function. These
counters can also be a list-style-type that is mapped to an existent counter-style. Currently we
don&apos;t handle TextDirection here, which breaks disclosure-closed style for the RTL case. Also here
we can delegate TextDirection processing by passing the element&apos;s direction to CSSCounterStyle::text().

* LayoutTests/TestExpectations:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemDisclosureClosed):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style disclosure-closed):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::updateContent):
* Source/WebCore/rendering/style/ListStyleType.cpp:
(WebCore::ListStyleType::isDisclosureClosed const): Deleted.
* Source/WebCore/rendering/style/ListStyleType.h:

Canonical link: <a href="https://commits.webkit.org/279404@main">https://commits.webkit.org/279404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/493c97dc7889b45d7056e5229e1660e646616a88

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56701 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/4147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3918 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/4147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55519 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46162 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/24441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3479 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2303 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46781 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58296 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52936 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28578 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46355 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11638 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65241 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29555 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12440 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -466,26 +466,27 @@ String RenderCounter::originalText() const
     };
     auto counterStyle = this->counterStyle();
-    String text = counterText(m_counter.listStyleType(), value, counterStyle.get());
+    auto textDirection = style().isLeftToRightDirection() ? TextDirection::LTR : TextDirection::RTL;
```
style().direction() should give you directly what you need instead of the ternary

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -317,7 +314,7 @@ void RenderListMarker::updateContent()
         auto counter = counterStyle();
         ASSERT(counter);
-        auto text = makeString(counter->prefix().text, counter->text(m_listItem->value()));
+        auto text = makeString(counter->prefix().text, counter->text(m_listItem->value(), style().isLeftToRightDirection() ? TextDirection::LTR : TextDirection::RTL));
```
Ditto style().direction() should give you what you need

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -127,10 +127,7 @@ auto RenderListMarker::textRun() const -> TextRunWithUnderlyingString
         if (style().isLeftToRightDirection())
             textForRun = m_textWithSuffix;
         else {
-            if (style().listStyleType().isDisclosureClosed())
```
Can we remove the isDisclosureClosed() method?

mdubet commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -236,6 +237,11 @@ static String counterForSystemCJK(int number, const std::array<UChar, 17>& table
 
+String CSSCounterStyle::counterForSystemDisclosureClosed(TextDirection textDirection)
+{
+    return textDirection == TextDirection::LTR ? span(blackRightPointingSmallTriangle) : span(blackLeftPointingSmallTriangle);
```
What is the `span` for ?

vitorroriz replied:
We don't have a String constructor for UChar IIRC

vitorroriz replied:
Yeah, we have removed String constructors that take raw char ptrs here: https://commits.webkit.org/276859@main

mdubet commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -466,26 +466,25 @@ String RenderCounter::originalText() const
             child = child->parent();
         while (CounterNode* parent = child->parent()) {
-            text = counterText(m_counter.listStyleType(), child->countInParent(), counterStyle.get())
+            text = counterText(m_counter.listStyleType(), child->countInParent(), counterStyle.get(), textDirection)
```
Why is this using textDirection but the line above is using style().direction ?

vitorroriz replied:
ops, I've messed this up when I've merged the 2 patches. Let me fix that.

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -127,10 +127,7 @@ auto RenderListMarker::textRun() const -> TextRunWithUnderlyingString
-                textForRun = span(blackLeftPointingSmallTriangle);
-            else
-                textForRun = makeString(reversed(StringView(m_textWithSuffix).substring(m_textWithoutSuffixLength)), m_textWithSuffix.left(m_textWithoutSuffixLength));
+            textForRun = makeString(reversed(StringView(m_textWithSuffix).substring(m_textWithoutSuffixLength)), m_textWithSuffix.left(m_textWithoutSuffixLength));
```
Now that the else branch is a single line, you can drop the brackets

nt1m commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -466,26 +466,25 @@ String RenderCounter::originalText() const
     int value = child->actsAsReset() ? child->value() : child->countInParent();
 
-    auto counterText = [](const ListStyleType& styleType, int value, CSSCounterStyle* counterStyle) {
+    auto counterText = [](const ListStyleType& styleType, int value, CSSCounterStyle* counterStyle, TextDirection textDirection) {
```
Can we make counterText capture this and only take in value as parameter?

vitorroriz committed:
[css-counter-styles] RenderListMarker, RenderCounter should use CSSCounterStyle::text() to handle TextDirection
https://bugs.webkit.org/show_bug.cgi?id=256153
rdar://109014745

Reviewed by Matthieu Dubet and Tim Nguyen.

Before we can address the bug, we need a small refactoring on how we handle
disclosure-closed counter style:
With this patch, we will handle disclosure-closed as one of the special complex systems,
such that we can consider RTL context when calculating its text value.

We fail css/css-counter-styles/counter-style-at-rule/disclosure-styles.html for 2 different
reasons. Therefore, we need to fix it in 2 different points:

1. RenderListMarker: This class is used for rendering list markers for the general case.
It has custom code for hadling TextDirection for counter styles due to disclosure-closed
style. We can now delegate this to CSSCounterStyle::text by passing TextDirection information.

The custom code currently checks if the counter-style has a "disclosure-closed" identifier but
that won't work for "extends" systems. When an author extends a counter-style, it can give any
identifier name to it, but it should still act as the counter-style it is extending.  Since this
is now handled by CSSCounterStyle itself, we no longer need the custom code.

2. RenderCounter: The class is used when rendering counters with the counter() function. These
counters can also be a list-style-type that is mapped to an existent counter-style. Currently we
don't handle TextDirection here, which breaks disclosure-closed style for the RTL case. Also here
we can delegate TextDirection processing by passing the element's direction to CSSCounterStyle::text().

* LayoutTests/TestExpectations:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemDisclosureClosed):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style disclosure-closed):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::updateContent):
* Source/WebCore/rendering/style/ListStyleType.cpp:
(WebCore::ListStyleType::isDisclosureClosed const): Deleted.
* Source/WebCore/rendering/style/ListStyleType.h:

Canonical link: https://commits.webkit.org/279404@main
